src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs

```diff
@@ -467,6 +467,15 @@ public BatchScope CreateBatchScope()
         }
     }
 
+    public async ValueTask<BatchScope> CreateBatchScopeAsync(CancellationToken cancellationToken)
+    {
+        using (await _gate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
+        {
+            _activeBatchScopes++;
+            return new BatchScope(this);
```

Comment:
Looks like we could thread the cancellation token through `BatchScope` to its `Dispose`/`DisposeAsync` methods, where those methods call `OnBatchScopeDisposedMaybeAsync` (which also awaits on the gate).

Reply:
Talked with Cyrus and we'll go ahead and give it a go

Reply:
Actually, I think I'm going to pass on this after looking into it a bit more. Previously, the Dispose method would always fully execute, regardless of whether cancellation had occurred. I don't think that should be changed.

Reply:
Fair enough. FWIW cancellation should only trigger from CPS projects when the project is being unloaded, at which point it's great to just bail immediately as everything will be torn down, but as a general mechanism it's safer to ensure a tidy disposal.